### PR TITLE
Feature/jibreen products endpoints

### PR DESCRIPTION
**Pull request by JbreenJboor1212**
### Product Management Endpoints

This module introduces CRUD (Create, Read, Update, Delete) endpoints for managing products in the e-commerce API. The implementation includes robust validation, error handling, and flexible search functionality.

#### 1. Create Product (`POST /`)
- Allows authorized admin users to create a new product.
- Ensures product names are unique.
- Returns product details, including `id`, `name`, `price`, `description`, `stock`, `isAvailable`, and `created_at`.

#### 2. Update Product (`PUT /{product_id}`)
- Allows authorized admin users to update a product by its UUID.
- Only updates the fields provided, leaving others unchanged.
- Ensures product name uniqueness during updates.
- Automatically updates the `updated_at` timestamp.

#### 3. Delete Product (`DELETE /{product_id}`)
- Allows authorized admin users to delete a product by its UUID.
- Returns `204 No Content` upon successful deletion.
- Raises an error if the product is not found.

#### 4. Get All Products (`GET /`)
- Retrieves a list of all available products, returning key information for each product.
- Returns an empty list if no products are available.

#### 5. Search Products (`GET /search`)
- Enables product search with flexible filters such as `name`, `min_price`, `max_price`, and `isAvailable` status.
- Supports sorting by `price` or `name`, with ascending or descending order.
- Includes pagination to manage large result sets.

#### 6. Get Product by ID (`GET /{product_id}`)
- Fetches detailed information about a specific product by its UUID.
- Provides product details such as `id`, `name`, `price`, `stock`, and timestamps.

### Validation and Error Handling
- Comprehensive validation for request inputs and proper error messages for various scenarios (e.g., duplicate product name, invalid UUID format).
- Uses appropriate HTTP status codes to signal success (`201 Created`, `200 OK`, `204 No Content`) and error conditions (`400 Bad Request`, `404 Not Found`).

### CODE REVIEW

**Review by Habeeb-Awawda — APPROVED** [review]

File: app/models.py [hunk 1]
```diff
@@ -157,3 +158,86 @@ class CreateStatusResponseModel(StatusModel):
...
+
+class UpdatedProductRequestModel(BaseModel):
+    name: Optional[str] =  Field(None, min_length=1, max_length=100)
+    price: Optional[Decimal] = Field(None,description="Product price", ge=Decimal('0.01'), decimal_places=2)
+    description: Optional[str] = Field(None, max_length=1000)
+    stock: Optional[int] = Field(None, ge=0)
+    isAvailable: Optional[bool] = Field(default=None)
+
+class UpdatedProductResponseModel(ProductBaseModel):
+    created_at: datetime
```
Comment (Mohammed-samaheen) [inline app/models.py:200]: ProductBaseModel doesn't have an id

File: app/api/routes/product.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            stock=product.stock,
+            isAvailable=product.isAvailable,
+            created_at=product.created_at,
+            updated_at=product.updated_at,
+        )
+        
+    except KeyError:
+        raise HTTPException(status_code=404, detail="Product with ID {product_id} not found.".format(product_id))
+    except ValueError:
+        raise HTTPException(status_code=400, detail="Product ID must be a valid UUID.")
```
Comment (Mohammed-samaheen) [inline app/api/routes/product.py]: no need for this one

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

File: app/api/routes/product.py (hunk 2 above)
Comment (JbreenJboor1212) [reply app/api/routes/product.py]: Done

File: app/models.py (hunk 1 above)
Comment (JbreenJboor1212) [reply app/models.py:200]: Done .

File: app/api/routes/product.py (hunk 2 above)
Comment (JbreenJboor1212) [reply app/api/routes/product.py]: Thank you for your feedback. I have a clearer perspective now and I'm working on avoiding the simple mistakes I made before and learning from them. The revisions have decreased, and the day errors were just due to lack of focus. I'll improve over time. @Mohammed-samaheen

**Review by Mohammed-samaheen — APPROVED** [review]

**Pull request merged**